arch/x86/kernel/slmodule.c

```diff
@@ -293,6 +293,174 @@ static void slaunch_intel_evtlog(void __iomem *txt)
 			SL_ERROR_TPM_INVALID_LOG20);
 }
 
+static void slaunch_tpm20_extend_event(struct tpm_chip *tpm, void __iomem *txt,
+				       struct tcg_pcr_event2_head *event)
+{
+	struct tpm_digest *digests;
+	u16 *alg_id_field = (u16 *)((u8 *)event +
+				    sizeof(struct tcg_pcr_event2_head));
+	u8 *dptr;
+	int ret;
+	u32 i, j;
+
+	digests = kcalloc(tpm->nr_allocated_banks, sizeof(struct tpm_digest),
+			  GFP_KERNEL);
+	if (!digests)
+		slaunch_txt_reset(txt,
+			"Failed to allocate array of digests\n",
+			SL_ERROR_GENERIC);
+
+	for (i = 0; i < tpm->nr_allocated_banks; i++)
+		digests[i].alg_id = tpm->allocated_banks[i].alg_id;
+
+
+	/* Early SL code ensured there was a max count of 2 digests */
+	for (i = 0; i < event->count; i++) {
+		dptr = (u8 *)alg_id_field + sizeof(u16);
+
+		for (j = 0; j < tpm->nr_allocated_banks; j++) {
+			if (digests[i].alg_id != *alg_id_field)
+				continue;
+
+			digests[i].alg_id = *alg_id_field;
+
+			switch (digests[0].alg_id) {
+			case TPM_ALG_SHA256:
+				memcpy(&digests[i].digest[0], dptr,
+				       SHA256_DIGEST_SIZE);
+				alg_id_field = (u16 *)((u8 *)alg_id_field +
+					SHA256_DIGEST_SIZE + sizeof(u16));
+				break;
+			case TPM_ALG_SHA1:
+				memcpy(&digests[i].digest[0], dptr,
+				       SHA1_DIGEST_SIZE);
+				alg_id_field = (u16 *)((u8 *)alg_id_field +
+					SHA1_DIGEST_SIZE + sizeof(u16));
+				break;
+			default:
+				;
```

Comment:
Thinking about it, it is possible the TPM has more banks than the ACM used (not sure if this is true but don't have evidence that it is not). So the default case should extend a zero hash and advance the alg_id_field.

Reply:
After looking the code over more closely, this is not correct. The alg_id_field is being walked correctly.